### PR TITLE
Address Google, Microsoft, Mozilla comments

### PR DESCRIPTION
**Pull request by azaroth42**
Not formal objections, but the substance of these organizations' comments deserves further treatment in the charter.

Preview: https://rawgit.com/json-ld/charter/ead95ff2a1b215dde6b8d5b79fddb138fa158759/index.html

Review requested also from @danbri (Do not merge without comment)
Thanks to and comments invited from @michaelchampion and @dbaron.

### CODE REVIEW

**Review by gkellogg — APPROVED** [review]

**Comment by iherman** [comment]
To make the comparison easier, I have created some more readable diff files:

* Diff to the version of the charter that was reviewed: https://www.w3.org/2018/05/diff_to_ac_reviewed.html
* Diff to the latest version on github that already included some (minor) changes based on the comments of other reviewers: https://www.w3.org/2018/05/diff_to_previous_changes.html

**Comment by iherman** [comment]
@azaroth42 two comments:

- Some of the changes are, I think, stylistic changes rather than substantial, mainly in the intro sections. I am not sure it is appropriate, at this stage, to make such changes (unless explicitly asked by some commenters). I would think those should be rolled back...
- The reference to the WoT has been removed; I believe the reference to the publishing WG should also be removed for similar reasons: there is currently no commitment in the Publ WG to use JSON-LD; it is just one of the options being considered

**Comment by BigBlueHat** [comment]
A few things...
#### 1. Scope
- Schema.org is not a format, but a vocabulary. So we should change this line:
   > "This is due largely to the adoption by schema.org as a recommended format.
- While schema.org may provide the largest numbers, it is not the only example of "success" or more specifically broad usage of JSON-LD as a format. ActivityStreams 2.0 and ActivityPub are used within Mastodon, PeerTube, and related decentralized social networking projects. There is not (currently) a good way to quantify the percentage of use that might be on the *public* Web. We should include a mention of that specification and usage also.
- JSON-LD is also actively used within organizations who's internal content is "beyond the reach" (and therefore quantification) by "public Web" search indexes. Highlighting some of the communities of use may be our best option.

#### 2. Deliverables
 -  The mention of 2.0 may still cause some concern wrt to backwards compatibility. Insisting that the approach will be *additive* may help here and that more complex features (if desired) will be clearly "opt-in" but that the architectural questions related to "how" we do that are exactly why we need this Working Group.

#### 3.1 W3C Groups
 - We should add the RDF.js Community Group to this list: https://www.w3.org/community/rdfjs/ Several library makers are consolidating their efforts and intend to collaborate on both implementing and iterating the current RDF.js specification: http://rdf.js.org/
 - We may also want to add the Open Annotation Community Group: https://www.w3.org/community/openannotation/ While the group is currently rather quiet, many of the members are active elsewhere and focused specifically on implementation (i.e. [Apache Annotator](http://annotator.apache.org/) (incubating)) and the [I Annotate](http://iannotate.org) conference community--represented primarily by [hypothes.is](http://hypothes.is/) and it's integration/interop partners.

**Comment by BigBlueHat** [comment]
There was also mention in one of the review emails that "an explicit phase" for engagement added to the charter.

Perhaps we simply need to highlight that the [JSON-LD Community Group](https://www.w3.org/community/json-ld/) has been in operation prior to this Working Group and that we intend to keep it and http://json-ld.org/ alive, well, and targeted at engaging not only tool (and spec) developers, but also developers interested in working with JSON-LD.

**Comment by azaroth42** [comment]
@iherman - The change to the introduction was to clarify the scope and rationale for the group, plus to remove the "in part" qualifier to schema.org which is less positive than it could be. That meant a little restructuring to the text. Also it removed the meaningless "grown enormously" ... grown from not existing is definitely grown. I can take out the bridge part but I think it addresses the need as raised by Mozilla and Microsoft.

The change to the second paragraph was to add clarity that we're not adding functionality to JSON-LD to express JSON, but to allow further idioms used in JSON to be part of the JSON-LD syntax.  It's a subtle distinction, and I can revert if you think it's a change too far.

The WoT coordination is still there, I just moved it down to de-emphasize it in the list, given the comment from Mozilla.

**Comment by azaroth42** [comment]
@BigBlueHat 

1.1. I reordered the sentence to make it clearer that the "it" which is a format is json-ld, not schema.org.
1.2,1.3. I agree, of course, that there are other uses (Web Annotations, IIIF, Linked.Art, LDP, SWWG specs, etc etc etc). However to Ivan's point, there wasn't a request to be more positive or even mention these.  Microsoft's point was that any TR that does not directly affect the document web should be carefully reviewed for acceptability ... which I do not agree with, but again, I don't think the semantic side of the house needs convincing.

2. Agree. I didn't want to arbitrarily take it out, but it was certainly something I considered.  What do others think? I would take it out if that doesn't prevent us from doing it if needed.

3. Groups -- I don't know what the policy is here.  My common sense based view is that this is the area where we call out particular groups that have expressed an interest in working together, not listing all possible groups that could be stakeholders.  That's what I was trying to cover in the extended intro to the list; that we'll target other stakeholder groups as needed and as possible over the course of the work to ensure the broadest consensus and awareness.

4.  Explicit phase is called "CR" :D

**Comment by danbri** [comment]
To be clear, the [Google response](https://lists.w3.org/Archives/Public/public-new-work/2018May/0000.html) to the charter proposal _was_ intended as a formal objection (which I anticipate can be addressed). Would it help if I reformatted https://lists.w3.org/Archives/Public/public-new-work/2018May/0000.html into a google doc with comments enabled maybe? (or any other doc sharing platform, I hear there are Annotation systems out there too....)

**Comment by iherman** [comment]
@azaroth42 

> @iherman - The change to the introduction was to clarify the scope and rationale for the group, plus to remove the "in part" qualifier to schema.org which is less positive than it could be. That meant a little restructuring to the text. Also it removed the meaningless "grown enormously" ... grown from not existing is definitely grown. I can take out the bridge part but I think it addresses the need as raised by Mozilla and Microsoft.
> The change to the second paragraph was to add clarity that we're not adding functionality to JSON-LD to express JSON, but to allow further idioms used in JSON to be part of the JSON-LD syntax. It's a subtle distinction, and I can revert if you think it's a change too far.

O.k., I understand your intention. Let us leave it (and hope nobody will complain:-)


> The WoT coordination is still there, I just moved it down to de-emphasize it in the list, given the comment from Mozilla.

I know. What I am saying that publishing is exactly in the same position, in this respect, as WoT, ie, it should be de-emphasized the same way.

(Edited, see https://github.com/json-ld/charter/pull/25#issuecomment-385886701)

**Comment by iherman** [comment]
> @azaroth42 
>> @BigBlueHat 
>> The mention of 2.0 may still cause some concern wrt to backwards compatibility. Insisting that the approach will be additive may help here and that more complex features (if desired) will be clearly "opt-in" but that the architectural questions related to "how" we do that are exactly why we need this Working Group.
>
> Agree. I didn't want to arbitrarily take it out, but it was certainly something I considered. What do others think? I would take it out if that doesn't prevent us from doing it if needed.

I agree with the removal. As @BigBlueHat says, 1.1 shows the additive nature of the work, and a 2.0, ie, a more radical change, should not be our goal.

> Groups -- I don't know what the policy is here. My common sense based view is that this is the area where we call out particular groups that have expressed an interest in working together, not listing all possible groups that could be stakeholders. 

I agree with @azaroth42 here. The coordination section is for groups with whom we would have explicit, well, coordination, primarily within W3C. It is clear that there are a number of other groups out there that we should keep in mind, but that is a different matter.

**Comment by iherman** [comment]
@azaroth42, I just realized something (sorry about the confusion), namely the the reference to the WoT WG did remain in the list of coordination groups, hence the confusion. Sorry.

My point is that, also based on Mozilla's comment, this list should include only W3C groups that explicitly work with JSON-LD. Past WG-s (like Annotation) does count because we cannot work with them any more:-), and groups that _consider_ using JSON-LD but have not yet formally decided should not be listed either. Ie, I would move WoT and Publ WG to the introductory section as examples for "Other Groups that express interest", but not listing them explicitly as part of the coordination.

**Comment by azaroth42** [comment]
@iherman @BigBlueHat New commit addresses the 2.0 and groups issues.

@danbri Does the [charter now](https://rawgit.com/json-ld/charter/mmg-comments/index.html) resolve the intended formal objection?

**Comment by danbri** [comment]
Can you summarize what you changed?

**Comment by azaroth42** [comment]
Of course :)

* Move Publishing and WoT to introduction as interested, not committed (per Ivan and Mozilla review)
* Remove "in part" for schema.org and add other WGs that have adopted JSON-LD to Introduction by way of comparison.
* Add notion of bridging need for semantics and need for usability to Intro
* Remove "legacy" and "supersede" and further clarify the backwards compatibility requirement
* Remove paragraph that mentions the 2.0 version number possibility
* Expand the discussion of the HTTP protocol sections to include to address your concerns of practical implementation at web-scale. 
* Add toolchain comparison to sit along side RDF serialization comparison, e.g. ShEX/SHACL and JSON-Schema. 
* Made more affirmative statement that schema.org CG will be regularly consulted, rather than should be at some point. 
* Push FPWD to August, given the above (Though September would make me happier, especially given summer vacations!)

**Comment by iherman** [comment]
@azaroth42, seeing the list of changes, and the extra work that those changes lead to (for good reasons, do not take me wrong), I wonder whether we should consider making the chartered length of the WG longer. It is currently chartered for a year and a half, I think making it two years, until June 2020, would give us more breathing space. It is fine to finish the WG earlier than chartered, but it is more difficult to extend the charter.

WDYT?

@danbri any opinion on that?

**Comment by azaroth42** [comment]
:+1: to extending the charter for 6 months.  The amount of interest in the WG is fantastic, but even with the additional items, the (anticipated, hoped for) extra voices in the discussions will make it a longer process than we had thought originally.  As you say, it's easier to finish early than to extend the charter.

**Comment by iherman** [comment]
The latest version can be previewed here:

https://rawgit.com/json-ld/charter/mmg-comments/index.html

**Comment by danbri** [comment]
Talking with @iherman at W3C AC meeting,

Here's a suggestion that makes some things a bit more explicit.
 
> "There are tensions in JSON-LD's design, between a) having an intuitive developer-friendly surface syntax, b) having the mapping into RDF's data model being handled by remote contexts c) avoiding the fetching of remote context mappings reveal (e.g. to intermediaries or remote servers) the topical content of JSON-LD data."

Rob previously added a sentence, and @iherman is looking at tweaks in this area e.g. mentioning  security and efficiency issues raised by handling @context files. Rob's was:

> "Note that the WG may decide, based on editorial and readability considerations, to extract any HTTP protocol related sections into a separate Recommendation and/or Working Group Note to provide best practices for ensuring practical implementation at web-scale, while respecting users' privacy, including notions of specific versioning to prevent the need for unnecessary dereferencing. "

My feeling is that JSON-LD's dual nature (JSON surface + RDF graph interpretation) is the core issue here; it is a very delicate balance, it has under-examined security/privacy aspects, and that acknowledging fixed or semi-fixed syntactic profiles may be worth exploring.

**Comment by iherman** [comment]
@azaroth42 @BigBlueHat, I have committed a version we essentially agree on with @danbri. (Look at section 2.1). Are you fine with this?

**Comment by iherman** [comment]
Refreshing the comment of @azaroth42 (https://github.com/json-ld/charter/pull/25#issuecomment-386763920) listing the latest changes on the charter:

* Move Publishing and WoT to introduction as interested, not committed (per Ivan and Mozilla review)
* Remove "in part" for schema.org and add other WGs that have adopted JSON-LD to Introduction by way of comparison.
* Add notion of bridging need for semantics and need for usability to Intro
* Remove "legacy" and "supersede" and further clarify the backwards compatibility requirement
* Remove paragraph that mentions the 2.0 version number possibility
* Add toolchain comparison to sit along side RDF serialization comparison, e.g. ShEX/SHACL and JSON-Schema.
* Made more affirmative statement that schema.org CG will be regularly consulted, rather than should be at some point.
* Push FPWD to August and extended the charter by 6 months (to altogether 2 years)
* Added a statement on considering security and efficiency (multiple fetch, etc) web-scale issues related to the current handling of `@context` files

Cc: @danbri

**Comment by msporny** [comment]
@danbri wrote:
> it is a very delicate balance, it has under-examined security/privacy aspects

We specifically avoided putting any language that outlined best practices wrt. JSON-LD contexts into the specification until we had enough real world experience to make such statements. I'd argue that we still don't want to put that sort of language into the core specification, but rather refer to an external document to do so.

We have quite a number of  workable solutions from a privacy/security standpoint that we're exploring in the Verifiable Credentials and Decentralized Identifiers work, namely:

 * Using content-addressing to refer to JSON-LD Contexts.
 * Using decentralized ledgers to host JSON-LD Contexts.
 * Using hard-coded application JSON-LD Contexts.
 * Using versioned JSON-LD Contexts that are frozen after a period of time.

**Comment by azaroth42** [comment]
I'm :+1: on the additions. I also agree with @msporny that a separate document is likely better than the core specification for many of the implementation level details.  This is called out in the current text that it might be a separate recommendation or note, but there's a great value in providing that guidance.

**Comment by danbri** [comment]
Consider also "This specification defines JSON-LD 1.1, a JSON-based format to serialize RDF 1.1 Graphs and Datasets as used for Linked Data." -> "This specification defines JSON-LD 1.1, a JSON-based format to serialize RDF 1.1 Graphs and Datasets as used for Linked Data and in Web content."

(I don't want people pointing to https://www.amazon.com/CafePress-Linked-Large-Coffee-White/dp/B010LII03Y as our scope...)

**Comment by lanthaler** [comment]
I would prefer the charter to be clearer about compatibility. The current draft says

> All changes must preserve backward compatibility for JSON-LD 1.0 documents. This means that, when processing existing JSON-LD documents, JSON-LD 1.1 processors generate the same expanded output, unless that output is subject to errata in JSON-LD 1.0 or is otherwise an unspecified implementation detail.

As it currently stands no JSON-LD 1.0 compliant processor will be able to process any 1.1 document. The charter should make that crystal clear which probably also means this work should be labeled as JSON-LD 2.0 and not 1.1. If that's not the desired goal, the charter should reflect that instead.

**Comment by azaroth42** [comment]
Agreed that JSON-LD 1.0 processors will not be able to process 1.1 contexts completely due to the new features. However, that doesn't seem like a reason to go to 2.0 to me, otherwise every change would be a new major version. Happy to make it clearer in the charter that ...
(a) we will try to make it possible to have 1.1 contexts that degrade to 1.0-only features if possible
(b) 1.0 processors will not understand new features added in 1.1 (though I would have expected that to be obvious)
... if those would resolve the last of the Google objections?

**Comment by iherman** [comment]
@danbri, @samuelweiler : I have added the term "privacy", alongside security and efficiency, to the list of concerns related to `@context` files.

**Comment by iherman** [comment]
@danbri we really must move on with this charter. Could you, please, give your approval to be able to merge this PR and move on to the next step?

@azaroth42 @BigBlueHat : in view of the time passing, there is no way the group can start 1st of June as planned. Should we move all the dates by two weeks?

**Comment by danbri** [comment]
Ivan, you know from our f2f discussions that this is awaiting your
colleagues' security/privacy -related attention. How is that going?

**Comment by BigBlueHat** [comment]
> @azaroth42 @BigBlueHat : in view of the time passing, there is no way the group can start 1st of June as planned. Should we move all the dates by two weeks?

👍 to postponing (sadly). I don't think we can avoid it at this point.

@danbri I'm 👍 on the current draft as it stands. Was there more you wanted said within the charter wrt to security and privacy considerations?

**Comment by iherman** [comment]
@danbri I hope to get a reply this wee end. If there is an agreement with him, I will merge this PR and move to the next step. Ok?

**Comment by danbri** [comment]
@iherman I'd like to see his reply

**Comment by iherman** [comment]
@danbri of course. I presume the answer will be here.

**Comment by danbri** [comment]
(It's also a long/vacation weekend in the US)

**Comment by samuelweiler** [comment]
I am content with the language around privacy in this PR. I might have said: "The WG will [attempt to?] mitigate...", but this language is fine.

I wonder if we should relax the backward-compatibility constraint in service of the above.  Should we leave the opportunity for a 1.0 version file to track the user through context file fetches (compare to web bugs) or should we allow a 1.1 (or 2.0) parser to mitigate those issues? 

Since I didn't see it mentioned above, here is a link to Google's comments: https://lists.w3.org/Archives/Public/public-new-work/2018May/0000.html

**Comment by danbri** [comment]
I am happy proceeding to a WG on this basis, and (TV Raman can confirm
formally by email) retract Google's 'Formal Objection'.

The backwards compatibility issue is complex, and it is not clear where the
levers of power reside. A single JSON-LD document, perhaps accompanied by a
media-type declaration (script tag or from elsewhere) may engage
vocabularies and context files which have varied positions w.r.t.
versioning and compatibility. I trust that the charter proposal has taken
on board enough of the considerations that come with the Web search
deployment style for JSON-LD that we can work these things through in a WG
setting. The answers are not clear, but the issues are now at least getting
proper attention.

**Comment by iherman** [comment]
Thank you @danbri and @samuelweiler; I will now proceed with the charter. And enjoy your long week-end!
